applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp

```diff
@@ -0,0 +1,180 @@
+        VariableUtils().SetNonHistoricalVariableToZero(VELOCITY_H1_SEMINORM, r_elements_array);
+
+        KRATOS_CATCH("");
+    }
+
+    void DivergenceProcess::ExecuteBeforeOutputStep()
+    {
+        KRATOS_TRY;
+
+        // Extract information
+        const std::size_t dimension = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+
+        // Check and set number of elements and check dimension
+        KRATOS_ERROR_IF(dimension < 2 || dimension > 3) << "Inconsinstent dimension to execute the process. Dimension =" << dimension << " but needed dimension = 2 or dimension = 3." << std::endl;
+        KRATOS_ERROR_IF(mrModelPart.NumberOfElements() == 0) << "The number of elements in the domain is zero. The process can not be applied."<< std::endl;
+        const unsigned int number_elements = mrModelPart.NumberOfElements();
+
+        // Auxiliar containers
+        GeometryData::ShapeFunctionsGradientsType DN_DX;
+
+        // Iterate over the elements
+        #pragma omp parallel for firstprivate(DN_DX)
+        for(int i_elem = 0; i_elem < static_cast<int>(number_elements); ++i_elem) {
+            auto it_elem = mrModelPart.ElementsBegin() + i_elem;
+            auto& r_geometry = it_elem->GetGeometry();
+
+            // Current geometry information
+            const std::size_t number_nodes_element = r_geometry.PointsNumber();
+
+            // Build values vectors of the velocity
+            Vector values_x(number_nodes_element);
+            Vector values_y(number_nodes_element);
+            Vector values_z(number_nodes_element);
+            for(int i_node=0; i_node < static_cast<int>(number_nodes_element); ++i_node){
+                const auto &r_velocity = r_geometry[i_node].FastGetSolutionStepValue(VELOCITY);
+                values_x[i_node] = r_velocity[0];
+                values_y[i_node] = r_velocity[1];
+                if (dimension == 3) {
+                    values_z[i_node] = r_velocity[2];
+                }
+            }
+
+            // Set integration points
+            const auto& r_integration_method = r_geometry.GetDefaultIntegrationMethod(); // Default is 0
+            const auto& r_integration_points = r_geometry.IntegrationPoints(r_integration_method);
+            const std::size_t number_of_integration_points = r_integration_points.size(); // Default is 1
+
+            // Initialize auxiliary local variables
+            double aux_divergence = 0;
+            double aux_velocity_seminorm = 0;
+
+            // Get gradient shape functions and detJ0 on integration points
+            Vector detJ0;
+            r_geometry.ShapeFunctionsIntegrationPointsGradients(DN_DX, detJ0, r_integration_method);
+
+            // Loop over integration points
+            for ( IndexType point_number = 0; point_number < number_of_integration_points; ++point_number ){
+
+                // Compute local gradient
+                Vector grad_x;
```

Comment:
You are allocating `grad_x`,`grad_y` and `grad_z` lots of times. You can  make it more efficient by doing this outside the main loop and passing them inside the `firstpravate` clause.

Reply:
Done, thanks for the hint!